book-content/chapters/15-designing-your-types.md

```diff
@@ -0,0 +1,1444 @@
+
+By setting `TMetadata`'s default value to an empty, we can essentially make `TMetadata` optional:
+
+```tsx
+type ResourceStatus<TContent, TMetadata = {}> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+```
+
+Now, we can create a `StreamingPlaylist` type without providing a `TMetadata` type argument:
+
+```tsx
+type StreamingPlaylist = ResourceStatus<{
+  id: number;
+  name: string;
+  tracks: string[];
+}>;
+```
+
+If we hover over it, we'll see that it's typed as expected, with `metadata` being an empty object:
+
+```tsx
+type StreamingPlaylist =
+  | {
+      status: "unavailable";
+      reason: string;
+    }
+  | {
+      status: "available";
+      content: {
+        id: number;
+        name: string;
+        tracks: string[];
+      };
+      metadata: {};
+    };
+```
+
+Defaults can help make your generic types more flexible and easier to use.
+
+### Type Parameter Constraints
+
+To set constraints on type parameters, we can use the `extends` keyword.
+
+We can force the `TMetadata` type parameter to be an object while still defaulting to an empty object:
+
+```tsx
+type ResourceStatus<TContent, TMetadata extends object = {}> = // ...
+```
+
+There's also an opportunity to provide a constraint for the `TContent` type parameter.
+
+Both of the `StreamingPlaylist` and `StreamingAlbum` types have an `id` property in their `content` objects. This would be a good candidate for a constraint.
+
+We can create a `HasId` type that enforces the presence of an `id` property:
+
+```tsx
+type HasId = {
+  id: number;
+};
+
+type ResourceStatus<TContent extends HasId, TMetadata extends object = {}> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+```
+
+With these changes in place, it is now required that the `TContent` type parameter must include an `id` property. The `TMetadata` type parameter is optional, but if it is provided it must be an object.
+
+When we try to create a type with `ResourceStatus` that doesn't have an `id` property, TypeScript will raise an error that tells us exactly what's wrong:
+
+```ts twoslash
+// @errors: 2344
+type HasId = {
+  id: number;
+};
+
+type ResourceStatus<TContent extends HasId, TMetadata extends object = {}> =
+  | {
+      status: "available";
+      content: TContent;
+      metadata: TMetadata;
+    }
+  | {
+      status: "unavailable";
+      reason: string;
+    };
+
+// ---cut---
+type StreamingPlaylist = ResourceStatus<
+  {
+    name: string;
+    tracks: string[];
+  },
+  {
+    creator: string;
+    artwork: string;
+    dateUpdated: Date;
+  }
+>;
+```
+
+Once the `id` property is added to the `TContent` type parameter, the error will go away.
+
+#### Constraints Describe Required Properties
```

Comment:
Since you could have a constraint like `string`, `null | undefined` or `(...args: never) => unknown`... it might be better to refrain from using "properties" here.
```suggestion
#### Constraints Describe Requirements
```